Sets Windows process to be PER_MONITOR_DPI_AWARE (2.1.4-stable) - #11760

bitbionic opened the pull request:
NOTE: this code is from the 2.1.4 release tag - I'm assuming it should merge well into the 2.1 branch and maybe 3.0 . This sets Windows process to be PER_MONITOR_DPI_AWARE addressing #11760  - This seems to resolve the issue for the ticket. I was able to test on 3 separate Windows boxes all on Windows 10 against NVIDIA GTX660m, NVIDIA 1070, and Intel HD Graphics 620. All seemed to perform well. I attempted to follow the same pattern of dll usage as QueryDpiForMonitor function. 

Hopefully this is helpful to merge - I needed it for my deployments but hope it gets rolled into production. Feel free to ping me if you have questions.

bitbionic committed:
Sets Windows process to be PER_MONITOR_DPI_AWARE addressing
https://github.com/godotengine/godot/issues/11760 - This seems to resolve
the issue for the ticket. I was able to test on 3 separate Windows boxes
against NVIDIA GTX660m, NVIDIA 1070, and Intel HD Graphics 620. All seemed
to perform well. I attempted to follow the same pattern of dll usage as
QueryDpiForMonitor function. NOTE: this code is from the 2.1.4 branchm
-- Amended to comply with style formatting (replaced spaces w/tabs)

bitbionic committed:
Sets Windows process to be PER_MONITOR_DPI_AWARE addressing https://github.com/godotengine/godot/issues/11760 - This seems to resolve the issue for the ticket. I was able to test on 3 separate Windows boxes against NVIDIA GTX660m, NVIDIA 1070, and Intel HD Graphics 620. All seemed to perform well. I attempted to follow the same pattern of dll usage as QueryDpiForMonitor function. NOTE: this code is from the 2.1.4 branch

Noshyaar commented:
The changes didn't comply with clang-format. Please use `clang-format -i <filename>` and `git commit --amend`.

bitbionic committed:
Merging for style changes to comply with style guidelines - Merge branch 'desktopscaling-11761-on-2.1.4-stable' of https://github.com/bitbionic/godot into desktopscaling-11760-on-2.1.4-stable

Sets the Windows process to be PER_MONITOR_DPI_AWARE addressing #11760 -
This seems to resolve the issue for the ticket. I was able to test on 3
separate Windows boxes against NVIDIA GTX660m, NVIDIA 1070, and Intel HD
Graphics 620. All seemed to perform well. I attempted to follow the same
pattern of dll usage as QueryDpiForMonitor function. NOTE: this code is
from the 2.1.4-stable tag

bitbionic commented:
Alright - I amended the style to use tabs instead of spaces. So I'm looking at the documentation on DPI aware code and it looks like I used a header from the Win 10 SDK which only supports back to Windows 8.1. I think to support Windows 7 I need to use the older call of [SetProcessDPIAware](https://msdn.microsoft.com/en-us/library/windows/desktop/ms633543(v=vs.85).aspx)  HOWEVER - according to the documentation it's not guaranteed to be forward compatible and is being deprecated. Not sure the best way to handle that. I'm going to try rewriting to use the old API and hopefully it works on Win 10.

bitbionic commented:
[DPITests.zip](https://github.com/godotengine/godot/files/1385796/DPITests.zip)

Using the old API SetProcessDPIAware vs SetProcessDPIAwareness "works" on Windows 10, but for some reason my graphics are getting dithered/quantized - They're pixelated. The behavior is really strange because when the window doesn't have focus (like me trying to grab a screenshot) the graphics are great/normal - but when the window has focus it's pixelated. I attached a video demonstrating the difference between the old API (SetProcessDPIAware) and the new (SetProcessDPIAwareness ). I think for my product I'm going to have to go with the newer code base. I'll wait until I hear back from the devs as to which version they want - this one won't compile on the CI box b/c of the header required for the new API.

bitbionic committed:
Should fix CI build issues and Win 10 SDK dependencies for #11760 and
Godot 3.0 DPI implementation. Rather than requiring the Windows 10 SDK
scaling header that was breaking the builds, this simply mimics the
enumeration that was being used. This has been tested on Windows 7 and up
and it seems to work as expected for dpi scaling.

donkeybonks commented:
Perhaps you should first look for the newer API using LoadLibrary and if it can not be found, load the older API.

bitbionic commented:
I ended up using the same code (almost) and @reduz did on 3.0 and I've been testing it all day on multiple Windows platforms. The viewport seems to be scaling appropriately now with it.

donkeybonks commented:
There's a odd edge case with a Retina MacBook running boot camp and a non-retina external monitor set as primary display.

I can test this next week (when my laptop comes back from repair) if you'd like .

bitbionic commented:
Haha - wow - I never would have thought of that edge case - not even sure how we'd address that if it's problematic but, here's hoping it's gonna work out fine :)

vnen commented:
Why can't you use the same code eb0f9651b3bb12b79db4ee4f8862a3be16a11ea5? The only real difference is that instead of `SHC_PROCESS_SYSTEM_DPI_AWARE` you're using `SHC_PROCESS_PER_MONITOR_DPI_AWARE`. So maybe that's a change that should be applied to master too.

In any case, the `printf`s are not needed, so you can remove them.

donkeybonks commented:
I meant to test this but haven’t bought the little dongles and such for the new MacBook to connect my monitors, lol.

aaronfranke commented:
What's the status of this PR? Is it good to merge? Is it still needed?

akien-mga commented:
Sorry for the delay, but to be honest, I'd prefer to see #11760 fixed (and tested!) on the master branch before including a fix into 2.1.

At least now I'm about to release 2.1.5-stable hopefully in coming weeks, so I'd likely wait for after that release to merge this and hopefully get some testing from users compiling the 2.1 branch from source or a future 2.1.6 beta.

akien-mga commented:
Actually as mentioned in https://github.com/godotengine/godot/pull/11305#issuecomment-405579300, I really prefer to stick to the workflow of having new features implemented in master first, tested there by those who compile the master branch daily, and then backported to stable branches when relevant/needed.

I'm aware that the implementation of the fix for #11760 would likely be different on the master branch due to its HiDPI support which is missing in 2.1, but once it's merged there, we can look into backporting the whole subset of the HiDPI support that would be needed to achieve the same result as this PR, while keeping code similar to the master branch and thus easier to maintain.
